### ISSUE TITLE
IgnoreIf Guard description is raised when PermitReentryIf Guard Fails

### ISSUE THREAD

**Issue by brcaswell** [dotnet-state-machine/stateless#422]
I'm getting the incorrect\unexpected `IgnoreIf` guard description when testing a `PermitReentryIf` guard failing (on a chess related program I'm in the initial stages on).

Here are brief snippets of both the state-machine configuration and the calling of methods on the type, which fire off the trigger with parameter.

###  State-Machine Configuration of `Square` for handling member `ChessPiece` instance changing.

```c#

_machine.Configure(SquareState.Occupied)
	.IgnoreIf(_occupyTigger, (piece) => piece.Identifier.Equals(ChessPiece.Identifier), "Same piece occupying square")
	.PermitReentryIf(_occupyTigger, (piece) => !piece.BelongsTo.Equals(ChessPiece.BelongsTo), "Cannot occupy square with another piece owned by the same player!")
	.OnEntryFrom(_occupyTigger, SetChessPiece).Permit(SquareStateTrigger.Unoccupy, SquareState.Unoccupied);

 ```
 
 ## Program Main()
 ```c# 

square.Occupy(firstWhitePiece);
square.Occupy(firstWhitePiece);
square.Occupy(secondWhitePiece); //guard related exception expected (but guard description is not correct)
square.Occupy(firstBlackPiece);

 ```
 
 I put together a dotnetfiddle to reproduce the issue.
 
## [Reproduction Link - dotnetfiddle][1]

  [1]: https://dotnetfiddle.net/linW5Y

**Comment by DeepakParamkusam** [comment]
Shorter code to reproduce the issue (on .NET Core 3.1, Stateless 5.1.8) -
```
static class Program
{
    private static void Main(string[] args)
    {
        var fsm = new StateMachine<State, Trigger>(State.A);
        fsm.OnUnhandledTrigger((_, t, c) => Console.WriteLine($"{string.Join(",", c ?? Enumerable.Empty<string>())} on trigger {t}"));
        fsm.OnTransitionCompleted(t => Console.WriteLine($"{t.Source} -> {t.Destination} on {t.Trigger}"));

        fsm.Configure(State.A)
            .Permit(Trigger.X, State.B);

        fsm.Configure(State.B)
            .IgnoreIf(Trigger.X, () => false, "IgnoreIf guard failed")
            .PermitReentryIf(Trigger.X, () => false, "PermitReentryTrigger guard failed");   

        fsm.Fire(Trigger.X);
        fsm.Fire(Trigger.X);

        Console.ReadKey(true);
    }

    public enum State { A, B }
    public enum Trigger { X }
```
Expected output = "PermitReentryTrigger guard failed on trigger X"
Actual output = "IgnoreIf guard failed on trigger X"

The bug is two-part:

1. If no valid transition is found, only the first TriggerBehaviourResult is returned by below method (containing only the first guard description)-
https://github.com/dotnet-state-machine/stateless/blob/c4650d9d9bf7256b25ac6fdb74a00166f747509f/src/Stateless/StateRepresentation.cs#L89-L92

2. IgnoreIf guard does not call OnUnhandledTrigger exception handler but still populates the UnmetGuardConditions ICollection when the guard fails (i.e when trigger is NOT ignored).

In the current issue, IgnoreIf is the first TriggerBehaviourResult with non-zero ICollection, causing IgnoreIf guard description to be displayed instead of PermitReentryIf guard description.  

@brcaswell As a workaround, put your IgnoreIf clause after the PermitReentryIf clause.

**Comment by HenningNT** [comment]
Should be resolved on 5.10.0.
There is a alpha version available, would you mind checking if the issue has been resolved?

**Comment by brcaswell** [comment]
I tested against "5.10.0-dev-542672901". 

![image](https://user-images.githubusercontent.com/967010/107121096-37592f80-6856-11eb-9fa6-24e16b309248.png)

Though it is an improvement to have the guard description of `PermitReentryIf`, and the specified description\handling of the linked PR is observed.  The issue actually remains it that `IgnoreIf` Guard Description should not be raised.

If the decision to handle this issue is actually to propagate all Guard Descriptions when any guard fails, I'm fine with closing this issue.
